File: haskell/src/FS.hs
-- The following may be needed for doctests.
{-# LANGUAGE MultiParamTypeClasses, FlexibleContexts, FlexibleInstances, ScopedTypeVariables #-}

module FS where
import           System.Directory
import           Control.Exception (SomeException, catch)
import           Control.Applicative
import           Control.Monad.Reader
import           Result (Result(..), success, failure, result)
import qualified Result as R
import           RelMonad
import           RelResult

import           System.IO.Unsafe
import
    Data.IORef


-- $setup
-- >>> :set -XFlexibleContexts -XScopedTypeVariables
        
data FS a = FS { runFS :: FilePath -> IO (Result a) }
            deriving (Show, Eq) -- Just for testing

instance Functor FS where
  fmap = liftM
  
instance Applicative FS where
  pure = return
  (<*>) = ap

instance Monad FS where
  return a = FS $ \_ -> return (success a)
  
  fs >>= f = FS $ \cwd -> runFS fs cwd >>= inner cwd
    where
      inner cwd (Success a) = runFS (f a) cwd
      inner _   (Failure s) = return (failure s)

-- Print via unsafePerformIO, just for testing!
instance Show a => Show (FilePath -> IO (Result a)) where
    show f = show (unsafePerformIO (f testPath))

-- Just for testing!
instance Eq a => Eq (FilePath -> IO (Result a)) where
   f == g =  unsafePerformIO (f testPath) == unsafePerformIO (g testPath)
                               
testPath :: FilePath                  
testPath = undefined

fsFailure :: String -> FS a             
fsFailure msg = FS (\ _ -> return (Failure msg))

fsIOSuccess :: IO a -> FS a
fsIOSuccess io = FS (\ _ -> io >>= return . Success)

fsToIO (FS f) = f testPath

-- | Test IO effects by creating an IORef, along with an FS that sets it, and a check.
refMkSetViaFSCheck :: (FS () -> FS a) -> Bool                
refMkSetViaFSCheck fsVia = unsafePerformIO $ do
                       r <- newIORef "oops"
                       resFS <- fsToIO $ fsVia (fsIOSuccess(writeIORef r "did-it"))
                       rStr <- readIORef r
                       return (rStr == "did-it")              
                              

--------------FS Error handling functions---------------------------------------
mapResult :: (Result a -> Result b) -> FS a -> FS b
mapResult f fs = FS $ \cwd -> f <$> (runFS fs cwd)

-- | Set the error message in a failure case. Useful for providing contextual information without
-- having to inspect result.
-- NB: This discards any existing message.
--
-- prop> setMessage "error" (fsFailure "other") == (fsFailure "error" :: FS String)
-- prop> setMessage "error" (return "other") == (return "other")
setMessage :: String -> FS a -> FS a
setMessage msg = mapResult (R.setMessage msg)

-- | Adds an additional error message. Useful for adding more context as the error goes up the stack.
-- The new message is prepended to any existing message.
--
-- prop> addMessage "error" (fsFailure "other") == (fsFailure "errorother" :: FS String)
-- prop> addMessage "error" (return "other") == (return "other")
addMessage :: String -> FS a -> FS a
addMessage msg = mapResult (R.addMessage msg)

-- | Runs the first operation. If it fails, runs the second operation. Useful for chaining optional operations.
-- Returns the error of `self` iff both `self` and `other` fail.
--
-- prop> (rFailure "error") `FS.or` (return "other") == (return "other")
-- prop> (return "can") `FS.or` (return "other") == (return "can")
-- prop> (rFailure "error") `FS.or` (rFailure "other") == (rFailure "other" :: FS String)
or :: FS a -> FS a -> FS a
or fs1 fs2 =  FS $ \cdw -> runFS fs1 cdw >>= result (const (runFS fs1 cdw)) (const (runFS fs2 cdw))

-- | Like "finally", but only performs the final action if there was an error.
-- If `action` fails that error is swallowed and only the initial error is returned.
--
-- prop> refMkSetViaFSCheck (\fsSet -> (fsFailure "cleanup!") `onException` fsSet)
-- prop> (fsFailure "cleanup!") `onException` (return "not") == (fsFailure "cleanup!")
onException :: FS a -> FS b -> FS a
onException  a sequel =  FS $ \cwd -> runFS a cwd >>= result (return . success) (\s -> (runFS sequel cwd) >> return (failure s))

-- | Ensures that the provided action is always run regardless of if `this` was successful.
-- If `self` was successful and `sequel` fails it returns the failure from `sequel`. Otherwise
-- the result of `self` is returned.
--
-- prop> refMkSetViaFSCheck (\fsSet -> (fsFailure "notok") `finally` fsSet)
-- prop> refMkSetViaFSCheck (\fsSet -> (return "ok") `finally` fsSet)
-- prop> (fsFailure "notok") `finally` (return "justdoit") == (fsFailure "notok")
-- prop> (return "ok") `finally` (return "justdoit") == (return "ok")
finally :: FS a -> FS b -> FS a
finally a sequel = FS $ \cwd -> runFS a cwd >>=
                             result (\x -> (runFS sequel cwd) >> return (success x))
                                    (\s -> (runFS sequel cwd) >> return (failure s))
                              
-- | Applies the "during" action, calling "after" regardless of whether there was an error.
-- All errors are rethrown. Generalizes try/finally.
--
-- prop> refMkSetViaFSCheck (\fsSet -> bracket (return "init") (\ _ -> fsFailure "notok") (\ _ -> fsSet))
-- prop> refMkSetViaFSCheck (\fsSet -> bracket (return "init") (\ _ -> return "ok") (\ _ -> fsSet))
bracket :: FS a -> (a -> FS b) -> (a -> FS c) -> FS c
bracket before after during = do
  a <- before
  c <- finally (during a) (after a)
  return c


--------------------------------------------------------------------------------

-- | List files without a nice error message
listFiles :: FS [FilePath]
listFiles = FS $ \cwd -> catch (success <$> getDirectoryContents cwd) handleResult

handleResult :: SomeException -> IO (Result a)
handleResult _ = return (failure "failed")

-- | List files with a nicer error message
ls :: FS [FilePath]
ls = setMessage "Invalid path" listFiles



---------------Relative monad instance for FS relative to Result----------------

instance RelMonad Result FS where
  retRel r = FS (const.return $ r)
  fs >%= f = FS $ \cwd ->
    let applied = (f <$> (runFS fs cwd)) -- x :: IO (Result (FS b))
    in applied >>= result (\s -> runFS s cwd) (return . Failure)

-- | Tests for "syntax" functions derived via  'RelMonad Result FS'
--
-- prop> rSetMessage "error" (rFailure "other") == (rFailure "error" :: FS String)
-- prop> rSetMessage "error" (return "other")   == (return "other" :: FS String)
--       
-- prop> rAddMessage "error" (rFailure "other") == (rFailure "errorother" :: FS String)
-- prop> rAddMessage "error" (return "other")   == (return "other" :: FS String)
--
-- prop> (rFailure "error") `rOr` (return "other")   == (return "other" :: FS String)
-- prop> (return "can") `rOr` (return "other")       == (return "can" :: FS String)
-- prop> (rFailure "error") `rOr` (rFailure "other") == (rFailure "other" :: FS String)
--
-- prop> refMkSetViaFSCheck (\fsSet -> (rFailure "cleanup!") `rOnException` fsSet :: FS String)
-- prop> (rFailure "cleanup!") `rOnException` (return "not") == (rFailure "cleanup!" :: FS String)
--
-- prop> refMkSetViaFSCheck (\fsSet -> (rFailure "notok") `rFinally` fsSet :: FS String)
-- prop> refMkSetViaFSCheck (\fsSet -> (return "ok") `rFinally` fsSet  :: FS String)
-- prop> (rFailure "notok") `rFinally` (return "justdoit") == (rFailure "notok" :: FS String)
-- prop> (return "ok") `rFinally` (return "justdoit")      == (return "ok" :: FS String)
--                              
-- prop> refMkSetViaFSCheck (\fsSet -> rBracket (return "init") (\ _ -> rFailure "notok" :: FS String) (\ _ -> fsSet))
-- prop> refMkSetViaFSCheck (\fsSet -> rBracket (return "init") (\ _ -> return "ok" :: FS String) (\ _ -> fsSet))
        

File: scala/src/main/scala/FS.scala
//   Copyright 2015 Commonwealth Bank of Australia
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

package lambdajam

import java.io.File

import lambdajam.FST.FST

import scalaz.Monad

case class FS[A](fst: FST[A]) {
  def runPath(s: String): Result[A] = runFS(new File(s))
  val runFS: File => Result[A] = fst.run
  import FST._

  def map[B](f: A => B): FS[B] = fst.map(f).toFS /* FS {
    runFS(_) map f
  }*/

  def flatMap[B](f: A => FS[B]): FS[B] = fst.flatMap(f.andThen(_.fst)).toFS /* FS {
    file => runFS(file).fold(a => f(a).runFS(file), error => Result.error(error))
  }*/

  /**
    * Set the error message in a failure case. Useful for providing contextual information without
    * having to inspect result.
    *
    * NB: This discards any existing message.
    */
  def setMessage(message: String): FS[A] = fst.tSetMessage(message).toFS /* FS {
    runFS(_).setMessage(message)
  }*/

  /**
    * Adds an additional error message. Useful for adding more context as the error goes up the stack.
    *
    * The new message is prepended to any existing message.
    */
  def addMessage(message: String): FS[A] = fst.tAddMessage(message).toFS /* FS {
    runFS(_).addMessage(message)
  }*/

  /**
    * Runs the first operation. If it fails, runs the second operation. Useful for chaining optional operations.
    *
    * Returns the error of `self` iff both `self` and `other` fail.
    */
  def or(other: FS[A]): FS[A] = fst.tOr(other.fst).toFS /* FS {
    file => runFS(file).or(other.runFS(file))
  }*/

  /**
    * Like "finally", but only performs the final action if there was an error.
    *
    * If `action` fails that error is swallowed and only the initial error is returned.
    */
  def onException[B](sequel: FS[B]): FS[A] = fst.tOnException(sequel.fst).toFS /* FS {
    file => runFS(file).fold(Result.ok, error => {
      sequel.runFS(file)
      Result.error(error)
    })
  }*/

  /**
    * Ensures that the provided action is always run regardless of if `this` was successful.
    * Generalizes "finally".
    *
    * If `self` was successful and `sequel` fails it returns the failure from `sequel`. Otherwise
    * the result of `self` is returned.
    */
  def ensure[B](sequel: FS[B]): FS[A] = fst.tEnsure(sequel.fst).toFS /* FS {
    file => runFS(file).fold(a => {
      sequel.runFS(file).fold(_ => Result.ok(a), Result.error )
    }, error => {
      sequel.runFS(file)
      Result.error(error)
    })
  } */

  /**
    * Applies the "during" action, calling "after" regardless of whether there was an error.
    *
    * All errors are rethrown. Generalizes try/finally.
    */
  def bracket[B, C](after: A => FS[B])(during: A => FS[C]): FS[C] = fst.tBracket(after.andThen(_.fst))(during.andThen(_.fst)).toFS /* FS {
    file => runFS(file).fold(a => {
      val fsc = during(a)
      val fsb = after(a)
      fsc.runFS(file).fold(c => {
        fsb.runFS(file).fold(_ => Result.ok(c), Result.error)
      }, Result.error)
    }, error => {
      Result.error[C](error)
    })
  }*/
}

object FS extends ToRelResultOps {
  def fs[A](a: => A): FS[A] = apply {
    _ => Result.safe(a)
  }

  def apply[A](r: File => Result[A]): FS[A] = FS(FST(r))

  /** Lists the files in a directory but doesn't have a nice error message. */
  def listFiles: FS[List[String]] = apply(f => Result.safeNull(f.list).map(_.toList))

  /** List files but with a nice error message. */
  def ls: FS[List[String]] = listFiles.setMessage("File is not a directory!")

  implicit def FSMonad: Monad[FS] = new Monad[FS] {
    def point[A](v: => A) = fs(v)
    def bind[A, B](a: FS[A])(f: A => FS[B]) = a flatMap f
  }


  /************** Relative Monad instance for FS relative to Result *************/

  import FST._

  implicit val relMonad: RelMonad[Result, FS] = new RelMonad[Result, FS] {
    /** Similar to a `Monad.point` but expects a `Result`. */
    def rPoint[A](v: => Result[A]): FS[A] = apply {
      _ => v
    }

    /** Similar to a `Monad.bind` but expects a `Result`. */
    def rBind[A, B](ma: FS[A])(f: Result[A] => Result[FS[B]]): FS[B] = ma.flatMap { a =>
      f(Result.ok(a)).fold(identity, error => MonadResult[FST].raiseError[B](error).toFS)
    }
  }

  /** List files but with a nice error message using RelResult functions. */
  def rLS: FS[List[String]] = FS.listFiles.rSetMessage("File is not a directory!")
}
